lib/internal/util/types.js

```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+const ReflectApply = Reflect.apply;
+
+// This function is borrowed from the function with the same name on V8 Extras'
+// `utils` object. V8 implements Reflect.apply very efficiently in conjunction
+// with the spread syntax, such that no additional special case is needed for
+// function calls w/o arguments.
+// Refs: https://github.com/v8/v8/blob/d6ead37d265d7215cf9c5f768f279e21bd170212/src/js/prologue.js#L152-L156
+function uncurryThis(func) {
```

Comment:
Have you checked the generated code for `isUint8Array`? I don't think we get the `Reflect.apply` optimization with V8 6.1 already.

Reply:
Yes I have:

```asm
--- Raw source ---
(value) {
  return TypedArrayProto_toStringTag(value) === 'Uint8Array';
}


--- Optimized code ---
optimization_id = 0
source_position = 1086
kind = OPTIMIZED_FUNCTION
name = isUint8Array
stack_slots = 4
compiler = turbofan
Instructions (size = 194)
0x367169e055c0     0  55             push rbp
0x367169e055c1     1  4889e5         REX.W movq rbp,rsp
0x367169e055c4     4  56             push rsi
0x367169e055c5     5  57             push rdi
0x367169e055c6     6  493ba5e80c0000 REX.W cmpq rsp,[r13+0xce8]
0x367169e055cd     d  0f8656000000   jna 0x367169e05629  <+0x69>
0x367169e055d3    13  48bf59b6c858e83c0000 REX.W movq rdi,0x3ce858c8b659    ;; object: 0x3ce858c8b659 <JSFunction b (sfi = 0x2fc5e3493879)>
0x367169e055dd    1d  488b7727       REX.W movq rsi,[rdi+0x27]
0x367169e055e1    21  488b5d10       REX.W movq rbx,[rbp+0x10]
0x367169e055e5    25  53             push rbx
0x367169e055e6    26  498b55a0       REX.W movq rdx,[r13-0x60]
0x367169e055ea    2a  33c0           xorl rax,rax
0x367169e055ec    2c  ff5737         call [rdi+0x37]
0x367169e055ef    2f  a801           test al,0x1
0x367169e055f1    31  0f844e000000   jz 0x367169e05645  <+0x85>
0x367169e055f7    37  488b58ff       REX.W movq rbx,[rax-0x1]
0x367169e055fb    3b  f6430bc0       testb [rbx+0xb],0xc0
0x367169e055ff    3f  0f8545000000   jnz 0x367169e0564a  <+0x8a>
0x367169e05605    45  48bb711549e3c52f0000 REX.W movq rbx,0x2fc5e3491571    ;; object: 0x2fc5e3491571 <String[10]: Uint8Array>
0x367169e0560f    4f  483bd8         REX.W cmpq rbx,rax
0x367169e05612    52  0f840b000000   jz 0x367169e05623  <+0x63>
0x367169e05618    58  498b45c0       REX.W movq rax,[r13-0x40]
0x367169e0561c    5c  488be5         REX.W movq rsp,rbp
0x367169e0561f    5f  5d             pop rbp
0x367169e05620    60  c21000         ret 0x10
0x367169e05623    63  498b45b8       REX.W movq rax,[r13-0x48]
0x367169e05627    67  ebf3           jmp 0x367169e0561c  <+0x5c>
0x367169e05629    69  48bb1067f50000000000 REX.W movq rbx,0xf56710
0x367169e05633    73  33c0           xorl rax,rax
0x367169e05635    75  488b75f8       REX.W movq rsi,[rbp-0x8]
0x367169e05639    79  e8a2efe7ff     call 0x367169c845e0     ;; code: STUB, CEntryStub, minor: 8
0x367169e0563e    7e  eb93           jmp 0x367169e055d3  <+0x13>
0x367169e05640    80  e8bbe9cfff     call 0x367169b04000     ;; deoptimization bailout 0
0x367169e05645    85  e8c0e9cfff     call 0x367169b0400a     ;; deoptimization bailout 1
0x367169e0564a    8a  e8c5e9cfff     call 0x367169b04014     ;; deoptimization bailout 2
0x367169e0564f    8f  e8cae9cfff     call 0x367169b0401e     ;; deoptimization bailout 3
0x367169e05654    94  90             nop
0x367169e05655    95  90             nop
...

Source positions:
 pc offset  position
        26       553
        69      1086

Inlined functions (count = 1)
 0x3ce858ce6719 <SharedFunctionInfo args>

Deoptimization Input Data (deopt points = 4)
 index  bytecode-offset  trampoline_pc     pc
     0               28             80     2f 
     1               15             85     NA 
     2               15             8a     NA 
     3                0             8f     7e 

Safepoints (size = 30)
0x367169e055ef    2f  0000 (sp -> fp)       0
0x367169e0563e    7e  0000 (sp -> fp)       3

RelocInfo (size = 11)
0x367169e055d5  embedded object  (0x3ce858c8b659 <JSFunction b (sfi = 0x2fc5e3493879)>)
0x367169e05607  embedded object  (0x2fc5e3491571 <String[10]: Uint8Array>)
0x367169e0563a  code target (STUB)  (0x367169c845e0)
0x367169e05641  runtime entry  (deoptimization bailout 0)
0x367169e05646  runtime entry  (deoptimization bailout 1)
0x367169e0564b  runtime entry  (deoptimization bailout 2)
0x367169e05650  runtime entry  (deoptimization bailout 3)

--- End code ---
```

Note: function `b()` is in fact `%TypedArray%.prototype.toStringTag`, and `args` is the name given to the wrapper function created in `uncurryThis`, which as you can tell is fully inlined.

Note 2: I don't really care if `Reflect.apply` is fully optimized for now, because regardless of that this function is much faster than the check we are currently using (https://github.com/nodejs/node/pull/15663#issue-261426498).

Reply:
Ok, with more recent V8 this will get even better (I also have a patch to inline `TypedArray.prototype[@@toStringTag]`). Nevertheless, why don't we just use `F.p.call` here and avoid `uncurryThis` completely?

Reply:
@bmeurer `Function.prototype.call` can be overwritten by userland code. I don't want that possibility for a type checking function.

Reply:
I see. Makes sense.